src/gather_classes.rs

```diff
@@ -0,0 +1,732 @@
+                Ordering::Greater => Ordering::Greater,
+                Ordering::Equal => entry_point_a.cmp(entry_point_b),
+            }
+        });
+        println!("\t(contract_hash, entry_point_selector)\tcount");
+        for (entry_point, count) in table_vec {
+            println!(
+                "\t({}, {})\t{}",
+                entry_point.contract_hash, entry_point.selector, count
+            );
+        }
+    }
+}
+
+/// Checks that the report corresponds to the block_num and tries to serializes it.
+fn try_serialize_block_report(
+    block_num: u64,
+    report: &ClassHashesReport,
+) -> Result<(), anyhow::Error> {
+    assert!(report.blocks.len() == 1 && report.blocks.contains(&block_num));
+    try_serialize(path_for_report(block_num), report).context(format!("block_num: {block_num}"))
+}
+
+fn generate_report_for_comparison_path<P>(path: P) -> Result<ClassHashesReport, anyhow::Error>
+where
+    P: AsRef<Path>,
+{
+    let path_str = path.as_ref().to_str().expect("failed to unwrap path");
+
+    let mut comparison_file = OpenOptions::new()
+        .read(true)
+        .open(path.as_ref())
+        .context(format!("Reading comparison file: '{}'", path_str,))
+        .inspect_err(|err| {
+            warn!("Error reading file: {err}");
+        })?;
+
+    let mut data = String::new();
+    let _data_len = comparison_file.read_to_string(&mut data)?;
+    let json_obj: Value = serde_json::from_str(data.as_str()).expect("unwrapping json_obj");
+
+    let json_block_num = try_get_block_num_from_json(&json_obj);
+    let path_block_num = try_get_block_num_from_path(path_str);
+    let block_num = val_or_err(path_block_num, json_block_num)?;
+
+    let mut report = ClassHashesReport::new();
+    update_report(&mut report, &json_obj, block_num)?;
+    Ok(report)
+}
+
+/// Attempts to deserialize report by getting the block_num from path.
+/// If that fails, deserializes the file to json_obj and tries to deserialize report by using the block number in the json_obj.
+/// If that fails, generates a new report from the comparison path.
+///
+/// Returns: Ok value: `(did_deserialize, report)` where `did_deserialize` is
+/// `true` when block report was deserialized from a file and
+/// `false` when a new block report was generated.
+fn gather_report_for_comparison_path<P>(path: P) -> Result<(bool, ClassHashesReport), anyhow::Error>
+where
+    P: AsRef<Path>,
+{
+    let path_str = path.as_ref().to_str().expect("failed to unwrap path");
+
+    let path_block_num = try_get_block_num_from_path(path_str);
+    if let Ok(block_num) = path_block_num {
+        // Attempt to deserialize using block_num from path.
+        let result =
+            try_deserialize(&path_for_report(block_num)).context(format!("block_num: {block_num}"));
+        if let Ok(block_report) = result {
+            return Ok((true, block_report));
+        }
+    };
+    // Either deserialization failed or block_num was unable to be derived from path.
```

Comment:
Why would block num won't be able to be derived from the path?

Reply:
Now making this assumption.